request.go

```diff
@@ -1,29 +1,90 @@
 package sarama
 
-type requestEncoder interface {
+import (
+	"fmt"
+)
+
+type requestHeader struct {
+	length        int32
+	key           int16
+	version       int16
+	correlationID int32
+	clientID      string
+}
+
+func (r *requestHeader) decode(pd packetDecoder) (err error) {
+	if r.length, err = pd.getInt32(); err != nil {
+		return err
+	}
+	if r.key, err = pd.getInt16(); err != nil {
+		return err
+	}
+	if r.version, err = pd.getInt16(); err != nil {
+		return err
+	}
+	if r.correlationID, err = pd.getInt32(); err != nil {
+		return err
+	}
+	r.clientID, err = pd.getString()
+	return nil
+}
+
+type request interface {
 	encoder
+	decoder
 	key() int16
 	version() int16
 }
 
-type request struct {
-	correlationID int32
-	id            string
-	body          requestEncoder
+func requestEncoder(req request, correlationID int32, clientID string) encodeFunc {
```

Comment:
I will update my PR with that

Sent from my Moto G
On Jul 10, 2015 11:34 AM, "Maxim Vladimirsky" <horkhe@gmail.com> wrote:

> Yes, that what I was thinking about too, when I said that I can do it.
>
> Sent from my Moto G
> On Jul 10, 2015 10:52 AM, "Evan Huus" <notifications@github.com> wrote:
>
>> In request.go
>> <https://github.com/Shopify/sarama/pull/478#discussion_r34373307>:
>>
>> >  	key() int16
>> >  	version() int16
>> >  }
>> >
>> > -type request struct {
>> > -	correlationID int32
>> > -	id            string
>> > -	body          requestEncoder
>> > +func requestEncoder(req request, correlationID int32, clientID string) encodeFunc {
>>
>> does something like
>> https://github.com/Shopify/sarama/compare/sample-request-body-refactor
>> make sense to you?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/Shopify/sarama/pull/478/files#r34373307>.
>>
>